File: src/ChaCha20EncryptDecrypt.php
<?php

class ChaCha20EncryptDecrypt
{
    private string $key;
    private string $nonce;

    public function __construct(string $key)
    {
        $this->key = sodium_crypto_generichash($key, '', 32);
        $this->nonce = random_bytes(SODIUM_CRYPTO_STREAM_NONCEBYTES);
    }

    public function encrypt(string $data): string
    {
        $ciphertext = sodium_crypto_stream_xor($data, $this->nonce, $this->key);
        return base64_encode($this->nonce . $ciphertext);
    }

    public function decrypt(string $data): string
    {
        $decoded = base64_decode($data);
        $nonce = substr($decoded, 0, SODIUM_CRYPTO_STREAM_NONCEBYTES);
        $ciphertext = substr($decoded, SODIUM_CRYPTO_STREAM_NONCEBYTES);
        return sodium_crypto_stream_xor($ciphertext, $nonce, $this->key);
    }
}

File: src/ChaCha20EncryptDecrypt.py
import base64
import hashlib
import pysodium

class XSalsa20EncryptDecrypt:
    def __init__(self, key: str):        
        self.key = hashlib.blake2b(key.encode('utf-8'), digest_size=32).digest()

    def encrypt(self, data: str) -> str:        
        nonce = pysodium.randombytes(pysodium.crypto_stream_NONCEBYTES)        
        ciphertext = pysodium.crypto_stream_xor(data.encode('utf-8'), len(data), nonce, self.key)        
        encrypted = nonce + ciphertext
        return base64.b64encode(encrypted).decode('utf-8')

    def decrypt(self, data: str) -> str:        
        decoded = base64.b64decode(data)        
        nonce = decoded[:pysodium.crypto_stream_NONCEBYTES]
        ciphertext = decoded[pysodium.crypto_stream_NONCEBYTES:]        
        decrypted = pysodium.crypto_stream_xor(ciphertext, len(ciphertext), nonce, self.key)
        return decrypted.decode('utf-8')